test/digitalTwin/utility.test.ts

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import { Utility } from "../../src/DigitalTwin/pnp/src/common/utility";
+import { Constants } from "../../src/DigitalTwin/pnp/src/common/constants";
+import { ModelType } from "../../src/DigitalTwin/pnp/src/deviceModel/deviceModelManager";
+
+const { readFile, writeJson, pathExists } = require("fs-extra");
+
+jest.mock("fs-extra");
+
+describe("Utility", () => {
+  const folder = "root";
+  const name = "test";
+  const modelId = "urn:model:id:1";
+  const replacement = new Map<string, string>();
+  replacement.set(Constants.DIGITAL_TWIN_ID_PLACEHOLDER, modelId);
+  replacement.set("{foo}", "bar");
+
+  test("create file from template", async () => {
+    const templatePath = "templatePath";
+    const filePath = "filePath";
+    const writeOptions = {
+      spaces: Constants.JSON_SPACE,
+      encoding: Constants.UTF8
+    };
+    const template = '{"@id": "{DigitalTwinIdentifier}"}';
+    readFile.mockResolvedValueOnce(template);
+    await Utility.createFileFromTemplate(templatePath, filePath, replacement);
+    expect(writeJson).toHaveBeenCalledWith(filePath, { "@id": modelId }, writeOptions);
+  });
+
+  test("replace all when multiple matched", () => {
+    const template = "{DigitalTwinIdentifier},{DigitalTwinIdentifier}";
+    expect(Utility.replaceAll(template, replacement)).toBe(modelId + "," + modelId);
+  });
+
+  test("replace all when not matched", () => {
+    const template = "{digitalTwinIdentifier}";
+    expect(Utility.replaceAll(template, replacement)).toBe(template);
+  });
+
+  test("replace all with multiple pattern", () => {
+    const template = "{DigitalTwinIdentifier},{foo}";
+    expect(Utility.replaceAll(template, replacement)).toBe(modelId + ",bar");
```

Comment:
>bar [](start = 71, length = 3)

Use a const variable here.

Reply:
Is this a common programming rule in TS to use a const var wherever a const string/object is needed? I have this question because the programming may look like this:
const str = "hello world";
console.log(str);
instead of:
console.log("hello world");

Isn't that needless? Forgive me that I am new to TS, :)

Reply:
If possible, I would like to make it as .toBe("urn:model:id:1,bar") to make it more readable/straightforward

Reply:
fixed